gpii/node_modules/settingsHandlers/src/RemoteFileSettingsHandler.js

```diff
@@ -0,0 +1,527 @@
+                            fs.unlink(source, remoteFileSettingsHandler.callbackToPromise(promise));
+                        }, promise.reject);
+                    } else {
+                        promise.reject(err);
+                    }
+                } else {
+                    promise.resolve();
+                }
+            });
+        }
+    } else {
+        fs.unlink(destination, function (err) {
+            if (err && err.code !== "ENOENT") {
+                promise.reject(err);
+            } else {
+                promise.resolve();
+            }
+        });
+    }
+    return promise;
+};
+
+/**
+ * Setter for the remote file settings handler.
+ *
+ * {
+ *  "settings": {
+ *    "word-ribbon": "file1"
+ *  },
+ *  "options": {
+ *    "settings": {
+ *      "word-ribbon": {
+ *        "path": "c:\\ ...",
+ *        "url": "https://example.com/"
+ *      },
+ *      "excel-ribbon": {
+ *        "path": "c:\\ ...",
+ *        "url": "https://example.com/"
+ *      }
+ *    }
+ *  }
+ * }
+ *
+ * @param {Object} payload The payload.
+ * @return {Promise} Resolves with the response.
+ */
+gpii.settingsHandlers.remoteFileSettingsHandler.setImpl = function (payload) {
+    var promiseTogo = fluid.promise();
+
+    var results = {};
+    var promises = [];
+    fluid.each(payload.settings, function (settingValue, key) {
+        if (settingValue) {
+            var value = settingValue.value || settingValue;
+
+            var settingPromise = fluid.promise();
+            promises.push(settingPromise);
+            settingPromise.then(null, function (reason) {
+                fluid.fail("removeFileSettingsHandler failed for setting '" + key + "' in payload ", payload, reason);
+            });
+
+            var options = payload.options.settings[key];
+
+            var remoteFileDownload = gpii.settingsHandlers.remoteFileDownload({
+                members: {
+                    invokeBefore: options.invokeBefore,
+                    invokeAfter: options.invokeAfter,
+                    cache: options.cache
+                }
+            });
+
+            var p;
+            if (settingValue.restore) {
+                p = remoteFileDownload.restoreFile(options.path, settingValue.stashPath, value);
+            } else {
+                p = remoteFileDownload.applyFile(options.path, options.url, value);
+            }
+
+            p.then(function (result) {
+                results[key] = {
+                    oldValue: {
+                        value: value,
+                        stashPath: result,
+                        restore: true
+                    },
+                    newValue: value
+                };
+                settingPromise.resolve();
+            }, function (reason) {
+                fluid.log("remoteFileDownload failed", reason);
+                results[key] = {failed: true};
+                settingPromise.resolve();
+            });
```

Comment:
Remember to add here as the final listener to settingPromise a call to remoteFileDownload.destroy()

Reply:
remembered.